# Data Quality Assessment

#### In this notebook, i will inspect all of the datasets, check for data noise, clean data, and make sure that the data is ready for the analysis and modeling phases.

### 1- The Customer Demographic Dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
demographic_data = pd.read_excel('Customer_demographic_data.xlsx')


In [3]:
demographic_data.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [4]:
demographic_data.shape

(4000, 13)

### We will check data types first.


In [5]:
demographic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
customer_id                            4000 non-null int64
first_name                             4000 non-null object
last_name                              3875 non-null object
gender                                 4000 non-null object
past_3_years_bike_related_purchases    4000 non-null int64
DOB                                    3913 non-null datetime64[ns]
job_title                              3494 non-null object
job_industry_category                  3344 non-null object
wealth_segment                         4000 non-null object
deceased_indicator                     4000 non-null object
default                                3698 non-null object
owns_car                               4000 non-null object
tenure                                 3913 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2), object(9)
memory usage: 406.4+ KB


#### As we see, there is no  problem here with data types.

### Next we will check for missing values

#### Counting missing values in each column:


In [6]:
missing_data = demographic_data.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("") 

customer_id
False    4000
Name: customer_id, dtype: int64

first_name
False    4000
Name: first_name, dtype: int64

last_name
False    3875
True      125
Name: last_name, dtype: int64

gender
False    4000
Name: gender, dtype: int64

past_3_years_bike_related_purchases
False    4000
Name: past_3_years_bike_related_purchases, dtype: int64

DOB
False    3913
True       87
Name: DOB, dtype: int64

job_title
False    3494
True      506
Name: job_title, dtype: int64

job_industry_category
False    3344
True      656
Name: job_industry_category, dtype: int64

wealth_segment
False    4000
Name: wealth_segment, dtype: int64

deceased_indicator
False    4000
Name: deceased_indicator, dtype: int64

default
False    3698
True      302
Name: default, dtype: int64

owns_car
False    4000
Name: owns_car, dtype: int64

tenure
False    3913
True       87
Name: tenure, dtype: int64



#### Showing the percentage of null values for each column:

In [7]:
demographic_data.isnull().sum() / demographic_data.shape[0]

customer_id                            0.00000
first_name                             0.00000
last_name                              0.03125
gender                                 0.00000
past_3_years_bike_related_purchases    0.00000
DOB                                    0.02175
job_title                              0.12650
job_industry_category                  0.16400
wealth_segment                         0.00000
deceased_indicator                     0.00000
default                                0.07550
owns_car                               0.00000
tenure                                 0.02175
dtype: float64

#### We have missing data in the following columns :

In [8]:
Missing_data_columns=demographic_data.isnull().any().reset_index()
for k in Missing_data_columns[Missing_data_columns[0]==True]['index']:
    print(k)

last_name
DOB
job_title
job_industry_category
default
tenure


#### Dealing with missing data:

In [9]:
# for the last name column, the last name feature is usually  used in modeling , so let's just fill the missing values with: "LN_unkown"
demographic_data["last_name"].replace(np.nan, "LN_unkown", inplace=True)

In [10]:
# for the DOB column, we will replace the missing values by the average of the column.
#first we'll calculate the mean of the column: 
DOB_mean= demographic_data["DOB"].mean(axis=0)
DOB_mean

Timestamp('1977-07-12 23:56:41.277792736')

In [11]:
# now we replace the missing values by the mean:
demographic_data["DOB"].replace(np.nan,"1977-08-12", inplace=True)
demographic_data["DOB"]= pd.to_datetime(demographic_data["DOB"])

In [12]:
demographic_data.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot  LN_unkown    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [13]:
# The title of the job won't matter for our future analysis as much as the industry of the job, 
#for the moment we will drop the whole column,if the column is needed, we could easily resotre and 
# fill the missing values by "uknown" or by the most common job title
demographic_data.drop('job_title', axis = 1,inplace=True)

In [14]:
demographic_data.job_industry_category.value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [15]:
# for the job_industry_category column , we will drop the rows with missing values 
demographic_data.dropna(subset=["job_industry_category"], axis=0, inplace=True)

# reset index, because we droped rows
demographic_data.reset_index(drop=True, inplace=True)  

In [16]:
# checking the column default:
demographic_data.default.value_counts()

1                                         95
100                                       94
-1                                        93
-100                                      82
ï½ï½¨(´âï½â©                             49
                                          ..
ð¾ ð ð ð ð ð ð ð                          24
/dev/null; touch /tmp/blns.fail ; echo    24
社æç§å­¸é¢èªå­¸ç ç©¶æ                    23
âªâªtest⪠                               22
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ          22
Name: default, Length: 90, dtype: int64

In [17]:
# the column default makes no sens, we will drop the whole column.
demographic_data.drop('default', axis = 1,inplace = True)

In [18]:
# for the "tenure" column, we will replace the missing values by the  mean of the column.
Tenure_mean= demographic_data["tenure"].mean(axis=0)
demographic_data["tenure"].replace(np.nan,Tenure_mean, inplace=True)

#### Let's check if we missed any missing values:

In [19]:
demographic_data.isnull().sum() / demographic_data.shape[0]

customer_id                            0.0
first_name                             0.0
last_name                              0.0
gender                                 0.0
past_3_years_bike_related_purchases    0.0
DOB                                    0.0
job_industry_category                  0.0
wealth_segment                         0.0
deceased_indicator                     0.0
owns_car                               0.0
tenure                                 0.0
dtype: float64

#### Missing values problem solved! this is the dataframe we're left with:

In [20]:
demographic_data.head()

customer_id first_name  last_name gender  \
0            1    Laraine  Medendorp      F   
1            2        Eli    Bockman   Male   
2            3      Arlin     Dearle   Male   
3            4     Talbot  LN_unkown   Male   
4            6       Curr  Duckhouse   Male   

   past_3_years_bike_related_purchases        DOB job_industry_category  \
0                                   93 1953-10-12                Health   
1                                   81 1980-12-16    Financial Services   
2                                   61 1954-01-20              Property   
3                                   33 1961-10-03                    IT   
4                                   35 1966-09-16                Retail   

   wealth_segment deceased_indicator owns_car  tenure  
0   Mass Customer                  N      Yes    11.0  
1   Mass Customer                  N      Yes    16.0  
2   Mass Customer                  N      Yes    15.0  
3   Mass Customer                  N       No     7.0  
4  High Net Worth                  N      Yes    13.0

In [21]:
demographic_data.shape

(3344, 11)

### Checking data accuracy 

#### we will check the values of each column:

In [22]:
demographic_data.gender.value_counts()

Female    1688
Male      1566
U           88
F            1
Femal        1
Name: gender, dtype: int64

In [23]:
# we will change F and Femal to Female:
demographic_data["gender"].replace(["F","Femal"], "Female", inplace = True)
demographic_data.gender.value_counts()

Female    1690
Male      1566
U           88
Name: gender, dtype: int64

In [24]:
demographic_data.job_industry_category.value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [25]:
demographic_data.wealth_segment.value_counts()

Mass Customer        1682
High Net Worth        851
Affluent Customer     811
Name: wealth_segment, dtype: int64

In [26]:
demographic_data.deceased_indicator.value_counts()

N    3342
Y       2
Name: deceased_indicator, dtype: int64

In [27]:
demographic_data.owns_car.value_counts()

Yes    1709
No     1635
Name: owns_car, dtype: int64

#### Checking if there are any duplicated rows in the dataframe:

In [28]:
demographic_data.duplicated().sum()

0

### we can do further preprocessing for this dataset , but it depends on the model we're opting to build with it , for the moment , we'll leave it like this 

#### Saving the clean data to a csv file for future analysis.

In [29]:
demographic_data.to_csv('Cleaned_Customer_demographic_data.csv')

### 2- The Transactions dataset

#### We will follow the steps for the previous dataset!

In [30]:
Transactions_data = pd.read_excel('Transactions_data.xlsx')

In [31]:
Transactions_data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [32]:
Transactions_data.shape

(20000, 13)

In [33]:
Transactions_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
transaction_id             20000 non-null int64
product_id                 20000 non-null int64
customer_id                20000 non-null int64
transaction_date           20000 non-null datetime64[ns]
online_order               19640 non-null float64
order_status               20000 non-null object
brand                      19803 non-null object
product_line               19803 non-null object
product_class              19803 non-null object
product_size               19803 non-null object
list_price                 20000 non-null float64
standard_cost              19803 non-null float64
product_first_sold_date    19803 non-null float64
dtypes: datetime64[ns](1), float64(4), int64(3), object(5)
memory usage: 2.0+ MB


#### Same as the previous dataset, the types are correct.

### Missing values:

In [34]:
missing_data = Transactions_data.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("") 

transaction_id
False    20000
Name: transaction_id, dtype: int64

product_id
False    20000
Name: product_id, dtype: int64

customer_id
False    20000
Name: customer_id, dtype: int64

transaction_date
False    20000
Name: transaction_date, dtype: int64

online_order
False    19640
True       360
Name: online_order, dtype: int64

order_status
False    20000
Name: order_status, dtype: int64

brand
False    19803
True       197
Name: brand, dtype: int64

product_line
False    19803
True       197
Name: product_line, dtype: int64

product_class
False    19803
True       197
Name: product_class, dtype: int64

product_size
False    19803
True       197
Name: product_size, dtype: int64

list_price
False    20000
Name: list_price, dtype: int64

standard_cost
False    19803
True       197
Name: standard_cost, dtype: int64

product_first_sold_date
False    19803
True       197
Name: product_first_sold_date, dtype: int64



In [35]:
Transactions_data.isnull().sum() / Transactions_data.shape[0]

transaction_id             0.00000
product_id                 0.00000
customer_id                0.00000
transaction_date           0.00000
online_order               0.01800
order_status               0.00000
brand                      0.00985
product_line               0.00985
product_class              0.00985
product_size               0.00985
list_price                 0.00000
standard_cost              0.00985
product_first_sold_date    0.00985
dtype: float64

#### We notice that there is a common precentage : "0.00985" , <br>this might mean that the missing data is common in some of the rows.

In [36]:
Transactions_data.online_order.value_counts()

1.0    9829
0.0    9811
Name: online_order, dtype: int64

In [37]:
# for online_order column, we will drop the rows with missing data:
Transactions_data.dropna(subset=["online_order"], axis=0, inplace=True)
# reset index, because we droped rows
Transactions_data.reset_index(drop=True, inplace=True)  

In [38]:
# for the brand column, we will replace the missing data with the most frequent brand 'Solex' :
Transactions_data['brand'].replace(np.nan,"Solex", inplace = True) 

In [39]:
Transactions_data.product_line.value_counts()

Standard    13920
Road         3894
Touring      1213
Mountain      418
Name: product_line, dtype: int64

In [40]:
# for the product line column, we will replace the missing data with the most frequent line 'Standard' :
Transactions_data['product_line'].replace(np.nan,"Standard", inplace = True) 

In [41]:
Transactions_data.product_class.value_counts()

medium    13587
high       2952
low        2906
Name: product_class, dtype: int64

In [42]:
# for the product class column, we will replace the missing data with the most frequent class 'medium' :
Transactions_data['product_class'].replace(np.nan,"medium", inplace = True) 

In [43]:
Transactions_data.product_size.value_counts()

medium    12767
large      3900
small      2778
Name: product_size, dtype: int64

In [44]:
# for the product size column, we will replace the missing data with the most frequent class 'medium' :
Transactions_data['product_size'].replace(np.nan,"medium", inplace = True) 

In [45]:
Transactions_data.standard_cost.value_counts()

388.920000    460
954.820000    389
53.620000     269
161.600000    232
260.140000    230
             ... 
151.960000    123
206.350000    111
270.299988      1
312.735016      1
667.400024      1
Name: standard_cost, Length: 103, dtype: int64

In [46]:
# for the standard_cost column, we will replace the missing values by the  mean of the column.
avg_cost= Transactions_data["standard_cost"].mean(axis=0)
Transactions_data["standard_cost"].replace(np.nan,avg_cost, inplace=True)

In [47]:
Transactions_data.product_first_sold_date.value_counts()

33879.0    230
41064.0    223
37823.0    221
39880.0    220
38216.0    217
          ... 
41848.0    168
42404.0    167
41922.0    163
37659.0    161
34586.0    159
Name: product_first_sold_date, Length: 100, dtype: int64

In [48]:
# for the product_first_sold_date column, we will replace the missing values by the  mean of the column.
avg_date= Transactions_data["product_first_sold_date"].mean(axis=0)
Transactions_data["product_first_sold_date"].replace(np.nan,avg_date, inplace=True)

In [49]:
#checking the missing values once again
Transactions_data.isnull().sum() / Transactions_data.shape[0]

transaction_id             0.0
product_id                 0.0
customer_id                0.0
transaction_date           0.0
online_order               0.0
order_status               0.0
brand                      0.0
product_line               0.0
product_class              0.0
product_size               0.0
list_price                 0.0
standard_cost              0.0
product_first_sold_date    0.0
dtype: float64

In [50]:
Transactions_data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [51]:
Transactions_data.shape

(19640, 13)

#### Now that we checked for the missing values, let's move to the next step!

### Checking data accuracy .

In [52]:
Transactions_data.order_status.value_counts()

Approved     19467
Cancelled      173
Name: order_status, dtype: int64

#### we checked the other features earlier!

#### Checking if there are any duplicated rows in the dataframe:

In [53]:
Transactions_data.duplicated().sum()

0

### As the previous data set, we could do more cleaning , but it depends on the use case , we'll leave it as it is for the moment.

In [54]:
# saving the cleaned data to a csv file.
Transactions_data.to_csv('Cleaned_Transactions_data.csv')

### 3- The New costumers dataset

In [55]:
New_customers_data = pd.read_excel('New_customers_data.xlsx')

In [56]:
New_customers_data.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.66      0.8250      1.03125     0.876563   
1                  11        0.72      0.7200      0.90000     0.765000   
2                   5        0.79      0.7900      0.79000     0.790000   
3                   1        1.09      1.3625      1.36250     1.362500   
4                   9        0.43      0.4300      0.53750     0.537500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [57]:
New_customers_data.shape

(1000, 23)

In [58]:
New_customers_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
first_name                             1000 non-null object
last_name                              971 non-null object
gender                                 1000 non-null object
past_3_years_bike_related_purchases    1000 non-null int64
DOB                                    983 non-null datetime64[ns]
job_title                              894 non-null object
job_industry_category                  835 non-null object
wealth_segment                         1000 non-null object
deceased_indicator                     1000 non-null object
owns_car                               1000 non-null object
tenure                                 1000 non-null int64
address                                1000 non-null object
postcode                               1000 non-null int64
state                                  1000 non-null object
country                                1000 non-null 

#### types are correct.

### Missing values:

In [59]:
New_customers_data.isnull().sum() / New_customers_data.shape[0]

first_name                             0.000
last_name                              0.029
gender                                 0.000
past_3_years_bike_related_purchases    0.000
DOB                                    0.017
job_title                              0.106
job_industry_category                  0.165
wealth_segment                         0.000
deceased_indicator                     0.000
owns_car                               0.000
tenure                                 0.000
address                                0.000
postcode                               0.000
state                                  0.000
country                                0.000
property_valuation                     0.000
Unnamed: 16                            0.000
Unnamed: 17                            0.000
Unnamed: 18                            0.000
Unnamed: 19                            0.000
Unnamed: 20                            0.000
Rank                                   0.000
Value     

In [60]:
# for the last name column we will  fill the missing values with: "LN_unkown"
New_customers_data["last_name"].replace(np.nan, "LN_unkown", inplace=True)

In [61]:
# for the column dob we'l do as the previous case :
DOB_mean= New_customers_data["DOB"].mean(axis=0)
DOB_mean

Timestamp('1971-04-20 14:05:14.954221768')

In [62]:
New_customers_data["DOB"].replace(np.nan,"1971-04-20", inplace=True)
New_customers_data["DOB"]= pd.to_datetime(New_customers_data["DOB"])

In [63]:
New_customers_data.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.66      0.8250      1.03125     0.876563   
1                  11        0.72      0.7200      0.90000     0.765000   
2                   5        0.79      0.7900      0.79000     0.790000   
3                   1        1.09      1.3625      1.36250     1.362500   
4                   9        0.43      0.4300      0.53750     0.537500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [64]:
# we will drop the column " job title", as it is not as relevant as the industry category:
New_customers_data.drop('job_title', axis = 1,inplace=True)

In [65]:
New_customers_data.job_industry_category.value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [66]:
New_customers_data.isnull().sum() / New_customers_data.shape[0]

first_name                             0.000
last_name                              0.000
gender                                 0.000
past_3_years_bike_related_purchases    0.000
DOB                                    0.000
job_industry_category                  0.165
wealth_segment                         0.000
deceased_indicator                     0.000
owns_car                               0.000
tenure                                 0.000
address                                0.000
postcode                               0.000
state                                  0.000
country                                0.000
property_valuation                     0.000
Unnamed: 16                            0.000
Unnamed: 17                            0.000
Unnamed: 18                            0.000
Unnamed: 19                            0.000
Unnamed: 20                            0.000
Rank                                   0.000
Value                                  0.000
dtype: flo

#### checking for duplicates:

In [67]:
New_customers_data.duplicated().sum()

0

In [68]:
#dropping unamed columns:
New_customers_data.drop(['Unnamed: 16','Unnamed: 17','Unnamed: 18','Unnamed: 19','Unnamed: 20'], axis = 1,inplace=True)

In [69]:
New_customers_data.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB job_industry_category     wealth_segment deceased_indicator  \
0 1957-07-12         Manufacturing      Mass Customer                  N   
1 1970-03-22              Property      Mass Customer                  N   
2 1974-08-28    Financial Services  Affluent Customer                  N   
3 1979-01-28         Manufacturing  Affluent Customer                  N   
4 1965-09-21    Financial Services  Affluent Customer                  N   

  owns_car  tenure              address  postcode state    country  \
0      Yes      14     45 Shopko Center      4500   QLD  Australia   
1       No      16    14 Mccormick Park      2113   NSW  Australia   
2       No      10  5 Colorado Crossing      3505   VIC  Australia   
3      Yes       5   207 Annamark Plaza      4814   QLD  Australia   
4       No      19    115 Montana Place      2093   NSW  Australia   

   property_valuation  Rank     Value  
0                   6     1  1.718750  
1                  11     1  1.718750  
2                   5     1  1.718750  
3                   1     4  1.703125  
4                   9     4  1.703125

In [70]:
New_customers_data.shape

(1000, 17)

### Checking data accuracy .

In [71]:
for k in New_customers_data.columns:
    print(k)

first_name
last_name
gender
past_3_years_bike_related_purchases
DOB
job_industry_category
wealth_segment
deceased_indicator
owns_car
tenure
address
postcode
state
country
property_valuation
Rank
Value


In [72]:
New_customers_data.gender.value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [73]:
New_customers_data.wealth_segment.value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [74]:
New_customers_data.deceased_indicator.value_counts()

N    1000
Name: deceased_indicator, dtype: int64

In [75]:
New_customers_data.owns_car.value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [76]:
New_customers_data.address.value_counts()

02 Roth Drive          1
64 Armistice Point     1
76 Bartelt Center      1
844 Forster Place      1
25805 Eagan Place      1
                      ..
4 Arapahoe Terrace     1
72 Mccormick Circle    1
92214 Spenser Road     1
5356 Sugar Plaza       1
265 Stephen Trail      1
Name: address, Length: 1000, dtype: int64

In [77]:
New_customers_data.postcode.value_counts()

2232    9
2145    9
2168    7
2750    7
3029    7
       ..
4408    1
2151    1
4159    1
3121    1
3051    1
Name: postcode, Length: 522, dtype: int64

In [78]:
New_customers_data.state.value_counts()

NSW    506
VIC    266
QLD    228
Name: state, dtype: int64

In [79]:
New_customers_data.country.value_counts()

Australia    1000
Name: country, dtype: int64

In [80]:
New_customers_data.property_valuation.value_counts()

9     173
8     162
7     138
10    116
6      70
11     62
5      57
4      53
3      51
12     46
2      42
1      30
Name: property_valuation, dtype: int64

In [81]:
New_customers_data.Rank.value_counts()

760     13
259     12
455      9
904      9
386      9
        ..
785      1
773      1
754      1
355      1
1000     1
Name: Rank, Length: 324, dtype: int64

In [82]:
New_customers_data.Value.value_counts()

0.637500    13
1.062500    12
1.237500     9
0.892500     9
0.945625     9
            ..
0.630000     1
0.340000     1
0.875500     1
0.867000     1
1.142187     1
Name: Value, Length: 324, dtype: int64

### We'll save the dataframe as csv:

In [83]:
New_customers_data.to_csv('Cleaned_New_customers_data.csv')

### 4- The costumers_adress dataset

In [84]:
Costumer_adress_data = pd.read_excel('Customer_adress_data.xlsx')

In [85]:
Costumer_adress_data.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [86]:
Costumer_adress_data.shape

(3999, 6)

In [87]:
Costumer_adress_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
customer_id           3999 non-null int64
address               3999 non-null object
postcode              3999 non-null int64
state                 3999 non-null object
country               3999 non-null object
property_valuation    3999 non-null int64
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


### Missing values:

In [88]:
Costumer_adress_data.isnull().sum() / Costumer_adress_data.shape[0]

customer_id           0.0
address               0.0
postcode              0.0
state                 0.0
country               0.0
property_valuation    0.0
dtype: float64

#### There are no missing values in this dataset !

#### Checking for duplicates:

In [89]:
Costumer_adress_data.duplicated().sum()

0

### Checking data accuracy:

In [90]:
Costumer_adress_data.country.value_counts()

Australia    3999
Name: country, dtype: int64

In [91]:
Costumer_adress_data.state.value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

### We'll leave this data set as it is , further changes could be applied when the analysis and model building phases start.

In [93]:
Costumer_adress_data.to_csv('Customers_adress_data.csv')

### With this final dataset, we end this Phase of preprocessing the data, further operations could be applied in future phases.